port-chooser/src/main/java/org/terracotta/utilities/test/net/NetStat.java

```diff
@@ -0,0 +1,1297 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terracotta.utilities.test.net;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terracotta.utilities.exec.Shell;
+import org.terracotta.utilities.test.runtime.Os;
+
+import java.io.IOException;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+import java.text.CharacterIterator;
+import java.text.StringCharacterIterator;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Stream;
+
+import static java.util.Objects.*;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Produces a network status collection similar to that obtained using {@code netstat}.
+ * This class is intended for diagnostic purposes; elevated privileges may be required
+ * for full status collection.
+ *
+ * <h3>Implementation Notes</h3>
+ * <ul>
+ *   <li>
+ *     For a Windows platform, {@code powershell} must be available via {@code PATH}.
+ *   </li>
+ *   <li>
+ *     For a Mac OS X platform, {@code nettop} and {@code ps} must be available via {@code PATH}.
+ *   </li>
+ *   <li>
+ *     For a Linux platform, {@code lsof} and {@code ps} must be available via {@code PATH}.  For
+ *     a complete port list, the effective user must be permitted to use {@code sudo} to execute
+ *     the {@code lsof} command <i>without a password.</i>
+ *   </li>
+ *   <li>
+ *     Not all connections observed using {@code netstat} are shown by {@code lsof} --
+ *     see <a href="https://raw.githubusercontent.com/lsof-org/lsof/master/00FAQ"><code>lsof</code> FAQ</a>
+ *     item 3.29.
+ *   </li>
+ * </ul>
+ */
+public class NetStat {
+  private static final Logger LOGGER = LoggerFactory.getLogger(NetStat.class);
+
+  /**
+   * Private niladic constructor to prevent instantiation.
+   */
+  private NetStat() {
+  }
+
+  /**
+   * Gets the list of busy ports on the current host.
+   * <p>
+   * Note that individual ports may appear more than once in the returned list.
+   * Depending on the source, a port may be listed for both IPv4 and IPv6 uses;
+   * for some applications (e.g. {@code sshd}), a port may be shared among multiple
+   * processes.
+   * @return the list of {@link BusyPort} instances representing the busy TCP ports
+   * @throws RuntimeException if the busy port information cannot be obtained
+   */
+  public static List<BusyPort> info() {
+    try {
+      return Platform.getPlatform().netstat();
+    } catch (Exception e) {
+      LOGGER.error("Failed to obtain TCP port information", e);
```

Comment:
Log and throw?

Reply:
In this case, I've reconsidered -- I don't think we're gaining much by logging here.  Removing it.